Python/chapter02/2.7 - Intersection/miguel_soln_2.7.py

```diff
@@ -0,0 +1,245 @@
+        return h
+
+    def append(self, ll: 'LinkedList') -> None:
+        self.tail.next = ll.head
+        self.size += ll.size
+        ll.head = None
+        ll.size = 0
+
+    def reverse(self) -> None:
+        """
+        Reverses this linked list in place
+        :return:
+        """
+        if self.head is None:
+            return
+        prev = self.head
+        self.tail = prev
+        curr = prev.next
+        self.tail.next = None
+        while curr is not None:
+            old_next = curr.next
+            curr.next = prev
+            prev = curr
+            curr = old_next
+        self.head = prev
+
+    def __repr__(self):
+        return self.__str__()
+
+    def __str__(self):
+        if self.head is None:
+            return '<empty>'
+        ll = []
+        n = self.head
+        while n.next is not None:
+            ll.append('{} -> '.format(n.data))
+            n = n.next
+        ll.append(str(n.data))
+        return ''.join(ll)
+
+    def __eq__(self, other: object):
+        if not isinstance(other, LinkedList):
+            return NotImplemented
+        a = self.head
+        b = other.head
+        while a is not None and b is not None:
+            if a.data != b.data:
+                return False
+            # otherwise, advance both pointers
+            a = a.next
+            b = b.next
+        return a is None and b is None
+
+
+def intersection(ll1: LinkedList, ll2: LinkedList) -> Optional[Node]:
+    """
+    This function will determine if ll1 and ll2 intersect.
+    The intersection is defined based on reference, not value.
+    Runtime:  O(n^2)
+    Space Complexity:  O(1)
+    :param ll1: first input linked list
+    :param ll2: second input linked list
+    :return: The intersecting node or None
+    """
+    n1 = ll1.head
+    while n1 is not None:
+        n2 = ll2.head
+        while n2 is not None:
+            if n1 is n2:
+                return n1
+            n2 = n2.next
+        n1 = n1.next
+    return None
+
+
+class SharedLLStructure(NamedTuple):
+    first_segment: LinkedList
+    second_segment: LinkedList
+    shared_node: Node
+    other_list: LinkedList
+
+
+class TestIntersection(unittest.TestCase):
+
+    def setUp(self):
+        shared_structures = [
+            SharedLLStructure(
+                LinkedList(1, 2, 3),
+                LinkedList(5, 10, 11),
+                Node(4),
+                LinkedList(6, 7, 8, 9)
+            ),
+            SharedLLStructure(
+                LinkedList(1, 2, 3, 4, 5),
+                LinkedList(7),
+                Node(6),
+                LinkedList(9, 8, 7),
+            ),
+            SharedLLStructure(
+                LinkedList(1, 2),
+                LinkedList(4, 5),
+                Node(3),
+                LinkedList(9, 8, 7, 6, 7, 6, 7, 9),
+            )
+        ]
+
+        self.intersection_test_cases = []
+
+        for ll1_seg1, ll1_seg2, shared_node, ll2 in shared_structures:
```

Comment:
now you don't need these abbreviated variables!

Reply:
updated